### DISCUSSION TITLE
Problem with LUA operator priority

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#312] [Q&A] [answered]
Hi,

I really have some trouble with LUA and its operator precedence with AND vs OR

I have the following function : 

```
	local Src4ConfigChange = function()
		if (n == "lblPrstUA" and CfgSrcType ~= 3) or (n == "lblPrstUA" and CfgSrcType ~= 2) or (n == "lblPrstUB" and CfgSrcType ~= 3) or (n == "lblPrstUB" and CfgSrcType ~= 2) or (n == "lblPrstUC" and CfgSrcType ~= 3) or (n == "lblPrstUD" and CfgSrcType ~= 3) then
			SelectButton() -- Call Select Button Press/Release
			ConfigButton() -- Call Select Config Press/Release
			panel:sendMidiMessageNow(CtrlrMidiMessage({0xC4, 0x34})) -- Config Program 52 (34h) Select 1U Psets	
		end
		panel:getLabel("lblProgramChanges"):setText("1 Unit Preset")
	end
```

so my initial statement is : 

`if (n == "lblPrstUA" and CfgSrcType ~= 3) or (n == "lblPrstUA" and CfgSrcType ~= 2) or (n == "lblPrstUB" and CfgSrcType ~= 3) or (n == "lblPrstUB" and CfgSrcType ~= 2) or (n == "lblPrstUC" and CfgSrcType ~= 3) or (n == "lblPrstUD" and CfgSrcType ~= 3) then`

even if n = lblPrstUA and the CfgSrcType = 3, the function goes through although it shouldn't since it doen't meet the requirements.

I just want to say with english words : 
if n == "lblPrstUA"		and		(CfgSrcType ~= 3 or ~= 2)
or
if n == "lblPrstUB"		and 		(CfgSrcType ~= 3 or ~= 2)
or 
if n == "lblPrstUC"		and 		(CfgSrcType ~= 3)
or 
if n == "lblPrstUD"		and 		 (CfgSrcType ~= 3)

I put some parenthesis to priorize properly but it doesn't work. I can't get it work unless I repeat the function 6 times to cover the 6 variations.

Do you know where I am wrong in my statement?

thanks for your help, I came across this OR vs AND priority that breaks everything and I always had to split the statement in several ones.

Damien

**Comment by damiensellier** [comment]
Alright I tried everything and undestood what was wrong. 
It comes from "is not equal to" in several statements so instead of filtering down it allows everything :) I replaced with "equals" only and it's filtering properly:

```
		if (n == "lblPrstUA" and (CfgSrcType == 0 or CfgSrcType == 1)) or (n == "lblPrstUB" and (CfgSrcType == 0 or CfgSrcType == 1)) or (n == "lblPrstUC" and CfgSrcType ~= 3) or (n == "lblPrstUD" and CfgSrcType ~= 3) then
			SelectButton() -- Call Select Button Press/Release
			ConfigButton() -- Call Select Config Press/Release
			panel:sendMidiMessageNow(CtrlrMidiMessage({0xC4, 0x34})) -- Config Program 52 (34h) Select 1U Psets	
		end
```

**Comment by dnaldoog** [comment]
If you want to cut down the <code>if  - or - and</code> soup, you could do the following, which is admittedly verbose, but scalable!

<pre>
local set = {
    lblPrstUA = true,
    lblPrstUB = true,
    lblPrstUC = true,
    lblPrstUD = true
}

local criteria = {
    lblPrstUA = true,
    lblPrstUB = true,
    lblPrstUC = false,
    lblPrstUD = false
}

function f(n, c)
    if set[n] then
        if criteria[n] then
        if c == 0 or c == 1 then
        console(n.." == 0 or ==1")
            return true
            end
            else -- criteria = false
        if c ~= 3 then
          console(n.."~=3")
            return true
        end
        end -- criteria true/false
    end -- set
    console("criteria not met")
    return false
end

if f(n, CfgSrcType) then
    SelectButton() -- Call Select Button Press/Release
    ConfigButton() -- Call Select Config Press/Release
    panel:sendMidiMessageNow(CtrlrMidiMessage({0xC4, 0x34})) 
     -- Config Program 52 (34h) Select 1U Psets	
end
</pre>

**Reply by damiensellier** [reply]
Thanks @dnaldoog It seems more complicated at first sight but with more than 6 variations, it's a very nice function for filtering on a bigger scale. I'll try it right away! thanks you John.

**Comment by dnaldoog** [comment, marked as answer]
Here is an even cooler way to do this and a few lines of code!

<pre>

myCallback = function(mod, value, source)
    n = L(mod:getName()) -- keep global
    c = panel:getModulatorByName("CfgSrcType"):getModulatorValue() -- keep global

    local f = assert(loadstring("if c " .. criteria[n] .. " then return true else return false end")) or false
    console(tostring(f()))
end

criteria = {
    lblPrstUA = " == 0 or c == 1",
    lblPrstUB = " == 0 or c == 1",
    lblPrstUC = " ~= 3",
    lblPrstUD = " ~= 3",
    lblPrstUE = " == 4",
    lblPrstUF = " == 1",
    lblPrstUG = " == 4 or c == 6"
}

</pre>

[loadstring logic_1_2_0_2021-05-28_15-56.zip](https://github.com/RomanKubiak/ctrlr/files/6558897/loadstring.logic_1_2_0_2021-05-28_15-56.zip)

**Reply by damiensellier** [reply]
This is one is even cleaner and adaptable! thanks John I marked it as the solution.
